### PR TITLE
Automatically enable async signal handling when possible

### PR DESCRIPTION
**Pull request by danog**

### CODE REVIEW

File: lib/Loop/NativeDriver.php [hunk 1]
```diff
@@ -43,10 +43,14 @@ class NativeDriver extends Driver
...
     private $signalHandling;
 
+    /** @var bool */
+    private $manualSignalHandling;
+
     public function __construct()
     {
         $this->timerQueue = new Internal\TimerQueue;
         $this->signalHandling = \extension_loaded("pcntl");
+        $this->manualSignalHandling = !(\PHP_MAJOR_VERSION >= 7 && \PHP_MINOR_VERSION >= 1);
```
Comment (kelunik) [inline lib/Loop/NativeDriver.php]: Please use `PHP_VERSION_ID < 70100` instead.

Comment (danog) [reply lib/Loop/NativeDriver.php]: Done

**Comment by danog** [comment]
> I don't like such a behavior change. Usually you don't need async signal handling in async applications, because they're already fairly non-blocking and `pcntl_signal_dispatch` gets called often enough.

I just wanted to implement this to reduce the (small, but still noticeable) overhead of calling pcntl_signal_dispatch at each tick; and it doesn't really hurt to reduce overhead by delegating some stuff to PHP, when possible and not too complex.

**Comment by trowski** [comment]
I believe we discussed this in the past and decided it was better to not enable async signal dispatching by default because it removed the guarantee that only a single watcher callback could be invoked at any given time. With async signal dispatch, a signal watcher may be invoked at *any* time, making it even more difficult to reason about execution flow. Since #264 was fixed, you can invoke `pcntl_async_signals` in your application if you want more immediate signal handling, of course with the caveat I mentioned before. We could check if async signal dispatching is enabled before calling `pcntl_signal_dispatch`, but then we're adding another function call to the normal case. Feel free to test, but I'd be surprised if there was a performance benefit.

**Comment by danog** [comment]
Instead of adding an additional function call, we could check if async signal handling is enabled when a new signal handler is set, or maybe add a new method to manually enable/disable signal handling (I prefer the first option).

**Comment by danog** [comment]
Done

**Comment by trowski** [comment]
IMO this is unnecessary. You're adding complexity to avoid a single function call per tick. `pcntl_dispatch_signals` checks a counter for pending signals and immediately returns if there are no pending signals. There is essentially no benefit to be had here.

`pcntl_async_signals` can be toggled at any time, so checking it only on construction or when adding a signal watcher is insufficient.

**Comment by danog** [comment]
While it is true that `pcntl_async_signals` can toggled at any time, it is usually toggled while setting up the signal handlers.

I don't see the point of even using `pcntl_async_signals` (except maybe for a more immediate call of the callback) if the AMPHP event loop already handles signals on its own: with the current unmodified event loop, `pcntl_async_signals` just adds more overall overhead, forcing the use of both PHP's internal signal scheduler and AMPHP's scheduler.

**Comment by kelunik** [comment]
Closing in favor of a more predictable execution order.

**Pull request closed without merging**